In [1]:
import pandas as pd
import numpy as np

user = pd.read_excel(r'/Users/huyennguyen/Documents/MDD/SQL/Data Files/UserReviewsClean43LIWC.xlsx')
user.head()

,url,idvscore,reviewer,dateP,Rev,thumbsUp,thumbsTot,WC,Analytic,Clout,...,Comma,Colon,SemiC,QMark,Exclam,Dash,Quote,Apostro,Parenth,OtherP
0,https://www.metacritic.com/movie/bronson,8,'Longbottom94',"'Apr 25, 2013'",'Many have dismissed this film for not explor...,2,2,113,94.05,32.91,...,4.42,0.0,0.0,0.0,0.0,0.88,0,1.77,0.00,0.0
1,https://www.metacritic.com/movie/bronson,9,'MartinB',"'Oct 13, 2009'",'Anyone who doesn t like this movie simply ju...,0,1,103,57.11,38.53,...,0.97,0.0,0.0,0.0,0.0,0.00,0,1.94,5.83,0.0
2,https://www.metacritic.com/movie/bronson,10,'Jaakko',"'Jul 19, 2012'",'Not sure what to think at this film at first...,1,1,143,75.73,66.27,...,0.70,0.0,0.0,0.0,0.0,0.00,0,1.40,0.00,1.4
3,https://www.metacritic.com/movie/bronson,6,'CapoR',"'Oct 13, 2009'",'Nicely portrayed but it lacks the elements t...,0,1,14,83.81,23.75,...,0.00,0.0,0.0,0.0,0.0,0.00,0,14.29,0.00,0.0
4,https://www.metacritic.com/movie/bronson,8,'OrwellB.',"'Oct 10, 2009'",'Bronson is more than entertainment. It is ar...,0,0,20,93.26,69.14,...,0.00,0.0,0.0,0.0,0.0,0.00,0,10.00,0.00,0.0


# 1. Checking Data Types

In [2]:
user.info()
user.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 319662 entries, 0 to 319661
Data columns (total 100 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   url           319662 non-null  object 
 1   idvscore      316258 non-null  object 
 2   reviewer      316255 non-null  object 
 3   dateP         316249 non-null  object 
 4   Rev           316249 non-null  object 
 5   thumbsUp      316082 non-null  object 
 6   thumbsTot     316086 non-null  object 
 7   WC            319662 non-null  int64  
 8   Analytic      319662 non-null  float64
 9   Clout         319662 non-null  float64
 10  Authentic     319662 non-null  float64
 11  Tone          319662 non-null  float64
 12  WPS           319662 non-null  float64
 13  Sixltr        319662 non-null  float64
 14  Dic           319662 non-null  float64
 15  function      319662 non-null  float64
 16  pronoun       319662 non-null  float64
 17  ppron         319662 non-null  float64
 18  i  

,WC,Analytic,Clout,Authentic,Tone,WPS,Sixltr,Dic,function,pronoun,...,Comma,Colon,SemiC,QMark,Exclam,Dash,Quote,Apostro,Parenth,OtherP
count,319662.000000,319662.000000,319662.000000,319662.000000,319662.000000,319662.000000,319662.000000,319662.000000,319662.000000,319662.000000,...,319662.000000,319662.000000,319662.000000,319662.000000,319662.000000,319662.000000,319662.0,319662.000000,319662.000000,319662.000000
mean,125.306871,68.983672,47.182937,25.878692,61.621293,21.697131,17.255578,78.369608,45.465104,10.525952,...,4.958649,0.182035,0.146803,0.369935,2.666544,4.498793,0.0,5.539397,0.598571,2.246406
std,143.712342,26.680480,23.478708,26.565498,38.349311,26.941899,9.329078,19.034795,13.617332,6.237834,...,29.430429,3.938785,35.247968,47.534426,178.741797,117.620962,0.0,14.964354,18.718212,88.437303
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.0,0.000000,0.000000,0.000000
25%,32.000000,53.280000,30.400000,4.050000,25.770000,11.500000,12.380000,77.160000,42.860000,6.670000,...,1.150000,0.000000,0.000000,0.000000,0.000000,0.000000,0.0,1.230000,0.000000,0.000000
50%,99.000000,76.620000,50.000000,16.480000,77.000000,16.950000,16.670000,82.650000,48.570000,10.530000,...,3.820000,0.000000,0.000000,0.000000,0.000000,0.000000,0.0,1.960000,0.000000,0.000000
75%,155.000000,91.050000,63.490000,39.990000,99.000000,24.200000,21.210000,87.370000,53.010000,14.290000,...,6.250000,0.000000,0.000000,0.000000,0.000000,0.670000,0.0,6.060000,0.000000,0.000000
max,7899.000000,99.000000,99.000000,99.000000,99.000000,6143.000000,100.000000,100.000000,100.000000,100.000000,...,5200.000000,1100.000000,19900.000000,22300.000000,67333.340000,16600.000000,0.0,200.000000,10200.000000,21400.000000


### idvscore (the review's score), thumbsUp and thumbsTot came back as object instead of numbers. A score column should be a number. Pandas falls back to object when it finds even one text value in a column, so some text got into these columns. That's the first clue that the scraper made some broken rows.

# Shape

319,662 rows x 100 columns (same as the expert file plus `thumbsUp` and `thumbsTot`).

# Duplicate Rows

In [3]:
print("Duplicate Rows:", user.duplicated().sum())

Duplicate Rows: 922


In [4]:
before = len(user)
user = user.drop_duplicates().reset_index(drop=True)
print(f"Removed {before - len(user)} rows; {len(user)} remain")

Removed 922 rows; 318740 remain


In [5]:
print("Duplicate Rows:", user.duplicated().sum())

Duplicate Rows: 0


# NaN Values

In [6]:
na = user.isna().sum()
na[na > 0]
#list of columns with NaN values

idvscore     2606
reviewer     2609
dateP        2615
Rev          2615
thumbsUp     2782
thumbsTot    2778
dtype: int64

In [7]:
user[user.isna().any(axis=1)]

,url,idvscore,reviewer,dateP,Rev,thumbsUp,thumbsTot,WC,Analytic,Clout,...,Comma,Colon,SemiC,QMark,Exclam,Dash,Quote,Apostro,Parenth,OtherP
29,https://www.metacritic.com/movie/a-screaming-man,NaN,NaN,NaN,NaN,NaN,NaN,0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0,0.0,0.0,0.0
30,https://www.metacritic.com/movie/lovers-of-hate,NaN,NaN,NaN,NaN,NaN,NaN,0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0,0.0,0.0,0.0
34,https://www.metacritic.com/movie/viva-riva!,NaN,NaN,NaN,NaN,NaN,NaN,0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0,0.0,0.0,0.0
130,https://www.metacritic.com/movie/september-11,NaN,NaN,NaN,NaN,NaN,NaN,0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0,0.0,0.0,0.0
134,https://www.metacritic.com/movie/catching-out,NaN,NaN,NaN,NaN,NaN,NaN,0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0,0.0,0.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
317674,https://www.metacritic.com/movie/domestic-viol...,NaN,NaN,NaN,NaN,NaN,NaN,0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0,0.0,0.0,0.0
317919,https://www.metacritic.com/movie/giorni-e-nuvole,NaN,NaN,NaN,NaN,NaN,NaN,0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0,0.0,0.0,0.0
317920,https://www.metacritic.com/movie/the-last-mist...,NaN,NaN,NaN,NaN,NaN,NaN,0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0,0.0,0.0,0.0
318186,https://www.metacritic.com/movie/iousa,NaN,NaN,NaN,NaN,NaN,NaN,0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0,0.0,0.0,0.0


In [8]:
user = user.dropna().reset_index(drop=True) #drops any row that has a NaN, these rows are broken, not just incomplete, and they're under 1% of the data. Fixing them by hand isn't worth it.
print("Real NaN values left:", user.isna().sum().sum())
print("Rows:", len(user))

#Order matters: this runs before the next section turns empty reviews into NaN. 
#If it ran after, those empty reviews would be deleted too, and we want to keep them.

Real NaN values left: 0
Rows: 315957


# Find Hidden Missing Values e.g "None"
### Pandas can't tell that the text "None" means missing, so it counts "None" as a real reviewer name or date

In [9]:
text_cols = user.select_dtypes(include=['object']).columns
placeholders = ['', 'nan', 'none', 'null', 'n/a', 'na', '-', '?']

for col in text_cols:
    s = user[col].astype(str).str.strip().str.strip("'\"").str.strip().str.lower()
    found = s[s.isin(placeholders)].value_counts()
    if len(found):
        print(col, found.to_dict())

reviewer {'na': 2, 'nan': 1, 'none': 1}
Rev {'': 1255, '?': 1}


In [10]:
s = user['reviewer'].str.strip().str.strip("'\"").str.strip().str.lower()
user.loc[s.isin(['nan', 'na', 'none']), 'reviewer'] = np.nan

s = user['Rev'].str.strip().str.strip("'\"").str.strip()
user.loc[s.isin(['', '?']), 'Rev'] = np.nan

In [11]:
user[['reviewer', 'dateP', 'Rev', 'thumbsUp', 'thumbsTot']].isna().sum()

reviewer        4
dateP           0
Rev          1256
thumbsUp        0
thumbsTot       0
dtype: int64

In [12]:
print("Total rows:", len(user))
(user[['Rev', 'reviewer']].isna().mean() * 100).round(1)
#In words, "For each of these columns, what percentage of cells are blank?"

Total rows: 315957


Rev         0.4
reviewer    0.0
dtype: float64

# Data Types

In [13]:
user['idvscore'] = user['idvscore'].astype(int)
user['idvscore'].describe()

#It was loaded as object because a few broken scraper rows had text in this column.
#Those rows were removed by dropna() earlier, so every value left is a number. 

count    315957.000000
mean          6.606472
std           3.185642
min           0.000000
25%           5.000000
50%           7.000000
75%           9.000000
max          10.000000
Name: idvscore, dtype: float64

### Thumbs

In [14]:
for col in ['thumbsUp', 'thumbsTot']:
    num = pd.to_numeric(user[col], errors='coerce')
    print(col, "text values:", (num.isna() & user[col].notna()).sum())

thumbsUp text values: 0
thumbsTot text values: 1


### info() showed the thumbs columns as object, which means at least one value isn't a number, and dropna() only removes empty cells, not text, so we count the text values before converting to see exactly what gets dropped.

In [15]:
for col in ['thumbsUp', 'thumbsTot']:
    user[col] = pd.to_numeric(user[col], errors='coerce')

user = user.dropna(subset=['thumbsUp', 'thumbsTot']).reset_index(drop=True)
user[['thumbsUp', 'thumbsTot']] = user[['thumbsUp', 'thumbsTot']].astype(int)

print("Rows:", len(user))
print("thumbsUp > thumbsTot:", (user['thumbsUp'] > user['thumbsTot']).sum())
user[['idvscore', 'thumbsUp', 'thumbsTot']].dtypes

Rows: 315956
thumbsUp > thumbsTot: 0


idvscore     int64
thumbsUp     int64
thumbsTot    int64
dtype: object

### thumbsTot had 1 value containing scraped text; that row was dropped (315,957 → 315,956). Both columns are now integers and thumbsUp never exceeds thumbsTot.

### Dates

In [16]:
user['dateP'].sample(10, random_state=1)
#This cell shows 10 random dates so you can see what format they're in before converting them.

220311     'Feb 13, 2017'
45914      'Apr 22, 2013'
204058     'Mar 27, 2013'
263882      'Aug 7, 2020'
261646     'Aug 24, 2013'
234076     'Dec 28, 2010'
309627     'Mar 27, 2006'
31427      'Jun 19, 2018'
76576      'Sep 12, 2019'
127435      'Feb 4, 2022'
Name: dateP, dtype: object

### Same quote marks as in the expert file, so we strip them before converting.

In [17]:
user['dateP'] = pd.to_datetime(
    user['dateP'].str.strip().str.strip("'"),
    format='%b %d, %Y'
)

In [18]:
print(user['dateP'].isna().sum()) #number of missing dates
print(user['dateP'].min(), user['dateP'].max())
user['dateP'].dtype

0
2001-01-26 00:00:00 2022-03-03 00:00:00


dtype('<M8[ns]')

### Reviewer and Rev still have the quote marks around them.
### Some reviews were cut off by the scraper, so they have an opening quote but no closing one. That's why we remove the first and last quote separately instead of only matching pairs like in the expert file.

In [19]:
for col in ['reviewer', 'Rev']:
    user[col] = (
        user[col]
        .str.strip()
        .str.strip("'\"")
        .str.replace(r"\\[rnt]", " ", regex=True)
        .str.replace(r"\s{2,}", " ", regex=True)
        .str.strip()
    )

In [20]:
for col in ['reviewer', 'Rev']:
    print(col, user[col].str.match(r"^['\"]").sum())

reviewer 0
Rev 0


### Extract Title

In [21]:
# Extract the title slug from the URL and turn dashes into spaces
user["title_from_url"] = (
    user["url"]
    .str.extract(r"/movie/([^/?#]+)", expand=False)
    .str.replace("-", " ", regex=False)
)

# Some slugs end with a "(YYYY)" year tag -> strip it off
user["title"] = (
    user["title_from_url"]
    .str.replace(r"\s*\(\d{4}\)$", "", regex=True)
    .str.strip()
)

# Drop the helper column
user = user.drop(columns=["title_from_url"])

# Normalize casing/spacing so titles compare and group consistently
user["title"] = user["title"].str.lower().str.strip()

print("Missing titles remaining:", user["title"].isna().sum())
print("Remaining mojibake in title:", user["title"].str.contains("©", na=False).sum())

Missing titles remaining: 0
Remaining mojibake in title: 0


In [22]:
# move 'title' right after 'reviewer'
cols = [c for c in user.columns if c != 'title']
cols.insert(cols.index('reviewer') + 1, 'title')
user = user[cols]

# rows with a reviewer first, empty reviewers at the bottom
empty = user['reviewer'].isna() | user['reviewer'].astype(str).str.strip().str.lower().isin(['', 'none', 'nan'])
user = user.sort_values('reviewer', key=lambda _: empty, kind='stable').reset_index(drop=True)

user[['reviewer', 'title']].head()

,reviewer,title
0,Longbottom94,bronson
1,MartinB,bronson
2,Jaakko,bronson
3,CapoR,bronson
4,OrwellB.,bronson


In [23]:
# surrogate key: unique sequential id for each review, placed as the first column
user.insert(0, 'id_user_review', range(1, len(user) + 1))

print("Unique ids:", user['id_user_review'].is_unique)
user.head(3)

Unique ids: True


,id_user_review,url,idvscore,reviewer,title,dateP,Rev,thumbsUp,thumbsTot,WC,...,Comma,Colon,SemiC,QMark,Exclam,Dash,Quote,Apostro,Parenth,OtherP
0,1,https://www.metacritic.com/movie/bronson,8,Longbottom94,bronson,2013-04-25,Many have dismissed this film for not explorin...,2,2,113,...,4.42,0.0,0.0,0.0,0.0,0.88,0,1.77,0.00,0.0
1,2,https://www.metacritic.com/movie/bronson,9,MartinB,bronson,2009-10-13,Anyone who doesn t like this movie simply just...,0,1,103,...,0.97,0.0,0.0,0.0,0.0,0.00,0,1.94,5.83,0.0
2,3,https://www.metacritic.com/movie/bronson,10,Jaakko,bronson,2012-07-19,Not sure what to think at this film at first. ...,1,1,143,...,0.70,0.0,0.0,0.0,0.0,0.00,0,1.40,0.00,1.4


# Save

In [ ]:
#with pd.ExcelWriter('user_cleaned.xlsx', datetime_format='yyyy-mm-dd') as writer:
#    user.to_excel(writer, sheet_name='user', index=False)

#print("Saved:", user.shape)

Saved: (315956, 102)


In [ ]:
#check = pd.read_excel('user_cleaned.xlsx')
#print("Shape:", check.shape)
#print("id_user_review unique:", check['id_user_review'].is_unique)
#print(check.dtypes[['id_user_review', 'idvscore', 'reviewer', 'dateP', 'Rev', 'thumbsUp', 'thumbsTot']])
#check[['reviewer', 'dateP', 'Rev', 'thumbsUp', 'thumbsTot']].isna().sum()

KeyboardInterrupt: 

### Save to Excel. About 99 reviews start with "=", which Excel treats as formulas, so they come back blank. We accept this loss (0.03% of reviews).